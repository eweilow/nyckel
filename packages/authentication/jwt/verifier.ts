import { verify } from "jsonwebtoken";
import jwksClient from "jwks-rsa";
import { DecodedJWT } from "./verify";

export class Verifier {
  constructor(private jwksUri: string, private issuer: string) {}

  readonly client = jwksClient({
    jwksUri: this.jwksUri
  });

  readonly getKey = (
    header: { kid: string },
    callback: (err: Error | null, signingKey?: string) => void
  ) => {
    this.client.getSigningKey(header.kid, (err, key) => {
      if (err != null) {
        return callback(err);
      }

      const signingKey = key.publicKey || key.rsaPublicKey;
      if (signingKey == null) {
        return callback(new Error("signingKey cannot be null or undefined"));
      }

      callback(null, signingKey);
    });
  };

  async verifyJWT(token: string) {
    return new Promise<DecodedJWT>((resolve, reject) => {
      verify(
        token,
        this.getKey as any,
        {
          issuer: this.issuer
        },
        (err, decoded) => {
          if (err != null) {
            return reject(err);
          }
          resolve(decoded as any);
        }
      );
    });
  }
}
