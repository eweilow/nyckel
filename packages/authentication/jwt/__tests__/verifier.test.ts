import { Verifier } from "../verifier";
import { verify } from "jsonwebtoken";
import jwksClient from "jwks-rsa";

jest.mock("jwks-rsa");
jest.mock("jsonwebtoken");

describe("Verifier", () => {
  it("creates a jwksClient correctly", () => {
    const returnValue = {
      returned: "yes"
    };
    ((jwksClient as any) as jest.Mock).mockReturnValueOnce(returnValue);

    const verifier = new Verifier("param:jwksUri", "param:issuer");
    expect((jwksClient as any) as jest.Mock).toBeCalledWith({
      jwksUri: "param:jwksUri"
    });
    expect(verifier.client).toBe(returnValue);
  });

  it("runs verifyJWT correctly", async () => {
    const decoded = { decodedToken: "yes" };
    (verify as jest.Mock).mockImplementationOnce((a, b, c, cb) =>
      cb(null, decoded)
    );

    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const result = await verifier.verifyJWT("token");
    expect(verify as jest.Mock).toBeCalledWith(
      "token",
      verifier.getKey,
      {
        issuer: "param:issuer"
      },
      expect.any(Function)
    );
    expect(result).toBe(decoded);
  });

  it("rejects verifyJWT correctly", async () => {
    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const errors = [new Error("hmm"), 3, "error"];

    for (const error of errors) {
      const decoded = { decodedToken: "yes" };
      (verify as jest.Mock).mockImplementationOnce((a, b, c, cb) =>
        cb(error, decoded)
      );

      const promise = verifier.verifyJWT("token");
      await expect(promise).rejects.toBe(error);
    }
  });

  it("runs getKey properly when getSigningKey returns publicKey", async () => {
    const getSigningKey = jest.fn((id, cb) => {
      cb(null, { publicKey: "publicKey" });
    });
    ((jwksClient as any) as jest.Mock).mockReturnValueOnce({
      getSigningKey
    });
    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const promise = new Promise((resolve, reject) => {
      verifier.getKey({ kid: "kid" }, (err, key) =>
        err ? reject(err) : resolve(key)
      );
    });
    expect(getSigningKey).toBeCalledWith("kid", expect.any(Function));
    await expect(promise).resolves.toBe("publicKey");
  });

  it("runs getKey properly when getSigningKey returns rsaPublicKey", async () => {
    const getSigningKey = jest.fn((id, cb) => {
      cb(null, { rsaPublicKey: "rsaPublicKey" });
    });
    ((jwksClient as any) as jest.Mock).mockReturnValueOnce({
      getSigningKey
    });
    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const promise = new Promise((resolve, reject) => {
      verifier.getKey({ kid: "kid" }, (err, key) =>
        err ? reject(err) : resolve(key)
      );
    });
    expect(getSigningKey).toBeCalledWith("kid", expect.any(Function));
    await expect(promise).resolves.toBe("rsaPublicKey");
  });

  it("runs getKey properly when getSigningKey returns publicKey and rsaPublicKey", async () => {
    const getSigningKey = jest.fn((id, cb) => {
      cb(null, { publicKey: "publicKey", rsaPublicKey: "rsaPublicKey" });
    });
    ((jwksClient as any) as jest.Mock).mockReturnValueOnce({
      getSigningKey
    });
    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const promise = new Promise((resolve, reject) => {
      verifier.getKey({ kid: "kid" }, (err, key) =>
        err ? reject(err) : resolve(key)
      );
    });
    expect(getSigningKey).toBeCalledWith("kid", expect.any(Function));
    await expect(promise).resolves.toBe("publicKey");
  });

  it("rejects getKey properly when getSigningKey returns error", async () => {
    const error = new Error("err");
    const getSigningKey = jest.fn((id, cb) => {
      cb(error, null);
    });
    ((jwksClient as any) as jest.Mock).mockReturnValueOnce({
      getSigningKey
    });
    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const promise = new Promise((resolve, reject) => {
      verifier.getKey({ kid: "kid" }, (err, key) =>
        err ? reject(err) : resolve(key)
      );
    });
    expect(getSigningKey).toBeCalledWith("kid", expect.any(Function));
    await expect(promise).rejects.toBe(error);
  });

  it("rejects getKey properly when getSigningKey returns no keys", async () => {
    const getSigningKey = jest.fn((id, cb) => {
      cb(null, {});
    });
    ((jwksClient as any) as jest.Mock).mockReturnValueOnce({
      getSigningKey
    });
    const verifier = new Verifier("param:jwksUri", "param:issuer");
    const promise = new Promise((resolve, reject) => {
      verifier.getKey({ kid: "kid" }, (err, key) =>
        err ? reject(err) : resolve(key)
      );
    });
    expect(getSigningKey).toBeCalledWith("kid", expect.any(Function));
    await expect(promise).rejects.toThrowErrorMatchingSnapshot();
  });
});
