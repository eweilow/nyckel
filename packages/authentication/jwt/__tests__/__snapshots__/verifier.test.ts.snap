// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verifier rejects getKey properly when getSigningKey returns no keys 1`] = `"[Nyckel] signingKey cannot be null or undefined"`;
